src/PostgREST/ApiRequest.hs

```diff
@@ -140,18 +141,19 @@ userApiRequest schema req reqBody =
 
   ApiRequest {
     iAction = action
-  , iRange  = if singular then singletonRange 0 else rangeRequested hdrs
   , iTarget = target
+  , iRange = M.insert "limit" (rangeIntersection headerRange urlRange) $
+      M.fromList [ (cs k, restrictRange (read <$> v) allRange) | (k,v) <- qParams, isJust v, endingIn ["limit"] k ]
```

Comment:
I see you imported `readMaybe` above, what about doing `(join $ readMaybe <$> v)` to be on the safe side?